doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb

```diff
@@ -0,0 +1,1152 @@
+    "    )\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "R3or9OgXrLT1"
+   },
+   "source": [
+    "We have already load the images from to a Ray dataset, which will partition the whole dataset and distribute the data blocks across the nodes in cluster. You will benefit from faster parallel pre-processing and data ingestion. Notice that the ResNet model was pretrained with hard-coded normalization values. We'll keep these numbers the same for fine-tuning, as shown in *data_transforms*. More details can be found [here](https://pytorch.org/hub/pytorch_vision_resnet/)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define a label preprocessor with {class}`BatchMapper <ray.data.preprocessors.BatchMapper>`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from ray.data.preprocessors import BatchMapper\n",
+    "\n",
+    "class_to_idx = {\"ants\": 0, \"bees\": 1}\n",
+    "\n",
+    "# 1. Map the image folder names to label ids\n",
+    "def map_labels(batch: np.ndarray) -> np.ndarray:\n",
+    "    batch[\"label\"] = np.vectorize(class_to_idx.__getitem__)(batch[\"class\"])\n",
+    "    batch.pop(\"class\")\n",
+    "    return batch\n",
+    "\n",
+    "\n",
+    "label_preprocessor = BatchMapper(fn=map_labels, batch_format=\"numpy\")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "GCc2hzkVrLT2"
+   },
+   "source": [
+    "```{note}\n",
+    "Note that **batch** here refers to the chunk of data for preprocessing, not the batch for model training. To learn more about writing functions for mapping batches, read [writing user-defined functions](transform_datasets_writing_udfs).\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, define an image preprocessor with {class}`TorchVisionPreprocessor <ray.data.preprocessors.TorchVisionPreprocessor>`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from ray.data.preprocessors import TorchVisionPreprocessor\n",
+    "\n",
+    "# 2. Convert input image to tensors\n",
+    "def to_tensor(batch: np.ndarray) -> torch.Tensor:\n",
+    "    tensor = torch.as_tensor(batch, dtype=torch.float)\n",
+    "    # (B, H, W, C) -> (B, C, H, W)\n",
+    "    tensor = tensor.permute(0, 3, 1, 2).contiguous()\n",
+    "    # [0., 255.] -> [0., 1.]\n",
+    "    tensor = tensor.div(255)\n",
+    "    return tensor\n",
+    "\n",
+    "\n",
+    "# Data augmentation and normalization for training\n",
+    "# Just normalization for validation\n",
+    "data_transforms = {\n",
+    "    \"train\": transforms.Compose(\n",
+    "        [\n",
+    "            transforms.Lambda(to_tensor),\n",
+    "            transforms.RandomResizedCrop(224),\n",
+    "            transforms.RandomHorizontalFlip(),\n",
```

Comment:
do these random transforms need to happen per epoch, or just once at the beginning?

I believe generally, these random augmentations are applied per-epoch.

Reply:
hmm that's indeed a problem. 

The context is: we have two datasets, "train" and "val", each has a unique preprocessor. The "train" preprocessor has to be applied per-epoch, the "val" preprocessor will not. Can you think of a way to achieve that?